### PR TITLE
feat(data): add indexed resumable multimodal loading and exact token budgets

### PR DESCRIPTION
**Pull request by pzelasko**
# feat(data): add indexed resumable multimodal loading and exact token budgets

> **Stack 2/7:** [1/7 Encoder foundations](https://github.com/NVIDIA-NeMo/Speech/pull/16192) → **Indexed/resumable data** → [3/7 Packed speech training](https://github.com/NVIDIA-NeMo/Speech/pull/16194) → [4/7 Portable export + target-only vLLM](https://github.com/NVIDIA-NeMo/Speech/pull/16195) → [5/7 MTP serving](https://github.com/NVIDIA-NeMo/Speech/pull/16196) → [6/7 DFlash serving](https://github.com/NVIDIA-NeMo/Speech/pull/16239) → [7/7 CCE dependency](https://github.com/NVIDIA-NeMo/Speech/pull/16241)

Base: `main`

Head: `codex/pr16158-rewrite/02-indexed-resumable-data`

Tip: `d21a02e4ccc60050c53b54862508ee135ea53af4`

## Summary

- Add multimodal idxpacks, WDS-v2 catalogs, ShareGPT routing, and portable lazy native-tar access.
- Preserve exact native-tar byte ranges and expected member names in deferred `shar_ptr` audio sources.
- Embed exact manifest-row-to-tar-member ordinal mappings inside backward-compatible v3 idxpacks.
- Keep v2 idxpacks and AIS/path-only loading backward compatible; v2 retains load-time name recovery.
- Add exact audio-token estimation and exact resumable packed sampling.
- Add validation, atomic publication, relocation/rebinding, provenance, and independent failure policies.
- Pin Lhotse to `2.0.0a6` for the reviewed fixed-array and strict lazy-pointer APIs.

## Reviewer guide

The review question is: can data be indexed, routed, validated, sampled, resumed, and budgeted exactly without depending on SALM tensor execution or serving?

This PR is logically independent of PR 1; it is mechanically stacked on PR 1 so all six branches form one review chain. Model-facing padding-free batch shaping stays in PR 3.

## Attribution

The reconstructed commits retain source-commit references and preserve the original authors and author-matching DCO sign-offs.

## Validation

- `git diff --check`: passed.
- Compilation: passed.
- Focused data tests: 558 passed, 7 skipped.
- Indexed/resumable regression suite: 153 passed.
- Published-Lhotse `2.0.0a6` route/index-pack suite: 159 passed at PR 2 and at the stack tip.
- AIS GetBatch tarred-adapter coverage: 9 passed with trusted source sampling-rate metadata in the fixture.
- `uv lock --check`: passed.
- Branch-scoped and complete-stack pre-commit: passed.
- All replayed PR-2 commits are patch-identical to their pre-rebase counterparts.

### CODE REVIEW

File: tests/collections/common/test_lhotse_sharegpt_missing_policy.py [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+    if not fault_tolerant_audio_loading:
+        with pytest.raises(expected_error):
+            list(build())
+        return
+
+    permissive = build()
+    assert len(permissive) == 3
+    assert [item.id for item in permissive] == ["sample-0", "sample-2"]
+    with pytest.raises(IndexError, match="not decodable"):
+        permissive[1]
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply tests/collections/common/test_lhotse_sharegpt_missing_policy.py:131]: This expression is intentionally evaluated inside pytest.raises to trigger the configured missing-data policy. Its effect is the asserted exception.

File: tests/collections/common/test_lhotse_indexed_partition.py [hunk 2]
```diff
@@ -384,6 +616,56 @@
...
+        conversations = list(adapter)
+        yielded_ids = {conversation.id for conversation in conversations}
+        assert len(adapter) == N_CUTS
+        assert len(conversations) == N_CUTS - 2
+        assert rows[0]["id"] not in yielded_ids
+        assert rows[1]["id"] not in yielded_ids
+        assert rows[2]["id"] in yielded_ids
+        assert rows[3]["id"] in yielded_ids
+        with pytest.raises(IndexError):
+            adapter[0]
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply tests/collections/common/test_lhotse_indexed_partition.py:833]: This expression is intentionally evaluated inside pytest.raises to trigger the indexed access error under test. Its effect is the asserted exception.

File: tests/collections/common/test_indexed_salm_strict_validation.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    requested = CutSet([first, second])
+
+    def partial(conversations, load_audio):
+        return torch.empty(0), torch.empty(0, dtype=torch.long), CutSet([first])
+
+    monkeypatch.setattr(salm_dataset, "collate_conversation_audio_fault_tolerant", partial)
+    strict = SALMDataset(_Tokenizer(), strict_audio_loading=True)
+
+    with pytest.raises(RuntimeError, match="dropped or reordered conversations"):
+        strict[requested]
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply tests/collections/common/test_indexed_salm_strict_validation.py:50]: This expression is intentionally evaluated inside pytest.raises to trigger and verify strict validation. Its effect is the asserted exception, so assigning the result would not improve the test.

File: nemo/utils/data_utils.py [hunk 4]
```diff
@@ -269,15 +271,22 @@
...
             if not os.path.isdir(local_dir):
                 os.makedirs(local_dir, exist_ok=True)
 
-            with open(local_path, 'wb') as f:
-                f.write(open_best(path).read(), num_retries=num_retries)
+            temp_path = None
+            try:
+                with tempfile.NamedTemporaryFile(dir=local_dir, prefix='.download-', delete=False) as f:
+                    temp_path = f.name
+                    f.write(open_best(path, num_retries=num_retries).read())
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/utils/data_utils.py]: Resolved in f2946ad8db5732c670db68f8b87c15e66907fbb6: the source reader now uses a context manager, with a regression assertion that it is closed after caching.

File: nemo/collections/common/data/lhotse/indexed_adapters.py [hunk 5]
```diff
@@ -222,6 +531,112 @@
...
+            self.data_path,
+            idx_path=idx_path,
+            metadata_path=metadata_path,
+        )
+        self._len = int(self.metadata["sample_count"])
+
+    def __len__(self) -> int:
+        return self._len
+
+    def __getitem__(self, idx: int) -> TarSampleBundle:
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/common/data/lhotse/indexed_adapters.py:572]: ValueError is deliberate here: this path reports malformed or undecodable indexed payload data, not a missing key or out-of-range index. Raising IndexError could also be misinterpreted as normal sequence termination by callers.

File: tests/collections/common/test_lhotse_sharegpt_missing_policy.py [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+        )
+
+    assert [item.id for item in build()] == ["sample-0", "sample-2"]
+    if not indexed:
+        return
+
+    direct = build()
+    assert len(direct) == 3
+    with pytest.raises(IndexError, match="not decodable"):
+        direct[1]
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply tests/collections/common/test_lhotse_sharegpt_missing_policy.py:307]: This expression is intentionally evaluated inside pytest.raises to trigger the configured missing-data policy. Its effect is the asserted exception.

File: nemo/collections/common/data/lhotse/dataloader.py [hunk 7]
```diff
@@ -1097,19 +1169,29 @@
...
+    if not use_bucketing:
+        if bucket_batch_size is not None:
             logging.info("Auto-enabling use_bucketing=True because bucket_batch_size is set.")
             config.use_bucketing = True
-        elif config.bucket_duration_bins is not None:
+            use_bucketing = True
+        elif bucket_duration_bins is not None:
             logging.info("Auto-enabling use_bucketing=True because bucket_duration_bins is set.")
             config.use_bucketing = True
+            use_bucketing = True
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/common/data/lhotse/dataloader.py]: Resolved in f2946ad8db5732c670db68f8b87c15e66907fbb6 by removing the dead post-configuration assignment.

File: tests/collections/common/test_lhotse_sharegpt_tar_routing.py [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+
+    # One range now contains two regular members, so it cannot be addressed by
+    # a (shard, local-member) route record.
+    idx_path.write_bytes(struct.pack("<2Q", members[0].start, tar_path.stat().st_size))
+    with pytest.raises(ValueError, match="exactly one regular member"):
+        validate_native_tar_member_index(tar_path, idx_path)
+
+
+def test_native_tar_range_validation_scales_linearly():
+    class CountingInt(int):
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply tests/collections/common/test_lhotse_sharegpt_tar_routing.py:505]: Intentional test instrumentation: CountingInt inherits complete int ordering behavior and overrides only selected operators so this test can count which comparisons the routing algorithm performs. No production API is affected.

File: nemo/collections/common/data/lhotse/dataloader.py [hunk 9]
```diff
@@ -1097,19 +1169,29 @@
...
+    batch_duration = config.get("batch_duration")
+    use_multimodal_sampling = bool(config.get("use_multimodal_sampling", False))
+    batch_tokens = config.get("batch_tokens")
+
+    if not use_bucketing:
+        if bucket_batch_size is not None:
             logging.info("Auto-enabling use_bucketing=True because bucket_batch_size is set.")
             config.use_bucketing = True
-        elif config.bucket_duration_bins is not None:
+            use_bucketing = True
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/common/data/lhotse/dataloader.py]: Resolved in f2946ad8db5732c670db68f8b87c15e66907fbb6 by removing the dead post-configuration assignment.

File: tests/collections/common/test_lhotse_sharegpt_tar_routing.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (pzelasko) [reply tests/collections/common/test_lhotse_sharegpt_tar_routing.py:505]: Intentional test instrumentation: CountingInt inherits int equality behavior and overrides only __ne__ to count that operation. Adding __eq__ would change what the test measures, so no code change is appropriate.

File: tests/collections/common/test_lhotse_indexed_partition.py [hunk 10]
```diff
@@ -148,6 +152,234 @@
...
+        cuts = list(adapter)
+        yielded_ids = {cut.custom["cut_id"] for cut in cuts}
+        assert len(adapter) == N_CUTS
+        assert len(cuts) == N_CUTS - 2
+        assert rows[0]["cut_id"] not in yielded_ids
+        assert rows[1]["cut_id"] not in yielded_ids
+        assert rows[2]["cut_id"] in yielded_ids
+        assert rows[3]["cut_id"] in yielded_ids
+        with pytest.raises(IndexError, match="not decodable"):
+            adapter[0]
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply tests/collections/common/test_lhotse_indexed_partition.py:495]: This expression is intentionally evaluated inside pytest.raises to trigger the indexed access error under test. Its effect is the asserted exception.

File: pyproject.toml [hunk 11]
```diff
@@ -474,6 +474,7 @@ deep-ep = ["setuptools"]
 # metadata is mirrored below so this repository controls the CUDA wheel index.
 [tool.uv.sources]
 nemo_automodel = { git = "https://github.com/NVIDIA-NeMo/Automodel.git", rev = "7b15c9bfbc620bec4dd3b819fce2e48726a7ddeb" }
+lhotse = { git = "https://github.com/lhotse-speech/lhotse.git", rev = "31912a04028b214988c949a5f7c815788409c478" }
```
Comment (KunalDhawan) [inline pyproject.toml]: Just confirming, does this commit match with the `lhotse==2.0.0a3` dependency used for ASR/audio/SpeechLM under the non-uv install path?

File: nemo/collections/common/data/lhotse/nemo_adapters.py [hunk 12]
```diff
@@ -902,77 +922,31 @@ def _build_indexed_url_cut(self, data: dict, manifest_path: str, tar_path: str)
...
-        try:
-            audio_bytes = self._tar_readers[sid].get(member_name)
-        except KeyError:
-            if self.skip_missing_manifest_entries:
-                return None
-            raise
-        return self._build_indexed_cut(data, audio_bytes, manifest_path, tar_path)
+        if self._indexed_entry_is_explicitly_skipped(data, manifest_path, tar_path):
+            return None
+        return self._build_indexed_url_cut(data, manifest_path, tar_path)
```
Comment (KunalDhawan) [inline nemo/collections/common/data/lhotse/nemo_adapters.py]: This now routes local indexed tarred NeMo data through the manifest-only archive-member `AudioSource` path. That loses the eager `soundfile.info()` metadata we used in `_build_indexed_cut`: for manifests without `sampling_rate`, we stamp every recording as 16 kHz/channels=[0] and compute `num_samples` from that, so non-16k tarred datasets may not be resampled correctly because later `resample(config.sample_rate)` sees the fabricated rate. It also moves missing/corrupt local tar-member failures into the PyTorch dataset layer, where most ASR/audio Lhotse datasets still instantiate `AudioSamples(fault_tolerant=True)` unconditionally. Can we keep eager metadata/strict validation for local non-AIS indexed tarred data unless the manifest provides trusted audio metadata, or plumb the strict policy through all downstream `AudioSamples` users and add a non-16k indexed-local-tar regression test?

File: nemo/collections/common/data/lhotse/sharegpt_tar_routing.py [hunk 13]
```diff
@@ -0,0 +1,995 @@
...
+        speaker = turn.get("from")
+        if not isinstance(speaker, str):
+            raise TypeError("ShareGPT conversation turn roles must be strings")
+        if speaker.lower() not in ("human", "user"):
+            continue
+        text = turn.get("value", "")
+        if not isinstance(text, str):
+            raise TypeError("ShareGPT conversation turn values must be strings")
+        placeholder_count += _count_audio_placeholders(text, audio_placeholders)
+    if len(paths) > 1 and placeholder_count > 1 and placeholder_count != len(paths):
```
Comment (KunalDhawan) [inline nemo/collections/common/data/lhotse/sharegpt_tar_routing.py:363]: edge case: this returns audio paths even when `placeholder_count == 0`. In collection mode that seals non-empty route records for the row, but `_ShareGPTConversationParser.transform()` will produce no audio turns for the same row and the runtime later raises `ShareGPT route row ... contains unused records`. That makes the route builder accept a manifest that direct/paired ShareGPT would effectively treat as text-only, then fail only during training. Should we either reject rows with audio fields but no consumable human/user placeholder here, or emit an empty route for them and make the runtime contract explicit?

**Review by KunalDhawan — DISMISSED** [review]

File: pyproject.toml (hunk 11 above)
Comment (pzelasko) [reply pyproject.toml]: It's a newer commit, I will release it later and adjust after we merge this monstrosity of a PR stack

File: nemo/collections/common/data/lhotse/sharegpt_tar_routing.py (hunk 13 above)
Comment (pzelasko) [reply nemo/collections/common/data/lhotse/sharegpt_tar_routing.py:363]: Resolved in 8e9ec7045f: ShareGPT routing counts only human/user placeholders and emits an empty route for text-only rows, including stale unused audio fields and literal assistant-side placeholder text. User placeholders still require a valid audio field, malformed path values still fail, and the existing cardinality rules remain enforced. Added unit and end-to-end collection tests.

File: nemo/collections/common/data/lhotse/nemo_adapters.py (hunk 12 above)
Comment (pzelasko) [reply nemo/collections/common/data/lhotse/nemo_adapters.py]: Resolved in 8e9ec7045f without eager audio-header I/O. Lazy NeMo tarred rows now take trusted per-row sampling_rate (or legacy sample_rate), or an explicit input_sampling_rate fallback distinct from the model target sample_rate; missing metadata fails before audio loading. The dead eager decoder was removed, the loader-level strict/tolerant policy is propagated into every ASR Lhotse AudioSamples instance, and regression coverage includes an 8 kHz indexed row that resamples correctly after selection while header/tar access is forbidden during lookup.

File: pyproject.toml (hunk 11 above)
Comment (pzelasko) [reply pyproject.toml]: Resolved in 8e9ec7045f: every NeMo dependency extra and uv.lock now use the released PyPI lhotse==2.0.0a4, and the temporary Git source override is removed.

**Review by KunalDhawan — DISMISSED** [review]

**Review by KunalDhawan — DISMISSED** [review]

**Review by KunalDhawan — DISMISSED** [review]

**Review by chtruong814 — DISMISSED** [review]
